File: TP1_material/python/src/pyramid.py
import OpenGL.GL as GL              # standard Python OpenGL wrapper
import numpy as np                  # all matrix manipulations & OpenGL args
import glfw

class Pyramid:

    def __init__(self, shader):
        self.shader = shader

        # pyramid position buffer, Numpy array of our 3D coordinates
        position = np.array(((-0.5, -0.5, -0.5), (-0.5, 0.5, -0.5), (0.5, 0.5, -0.5), (0.5, -0.5, -0.5), (0, 0, 0.5)),'f')

        indices = np.array(((0, 1, 2), (0, 2, 3), (0, 1, 4), (1, 2, 4), (2, 3, 4), (0, 3, 4)), dtype=np.int8)

        self.glid = GL.glGenVertexArrays(1)  # create OpenGL vertex array id
        GL.glBindVertexArray(self.glid)      # activate to receive state below
        #self.buffers = [GL.glGenBuffers(2)]  # create buffer for position attrib
        self.buffers = GL.glGenBuffers(2)	 # if n > 1, use this instead
        # GL.glGenBuffers(n) with n > 1 directly returns a list and not an index

        # create position attribute, send to GPU, declare type & per-vertex size
        loc = GL.glGetAttribLocation(shader.glid, 'position')
        GL.glEnableVertexAttribArray(loc)    # assign to position attribute
        GL.glBindBuffer(GL.GL_ARRAY_BUFFER, self.buffers[0])
        GL.glBufferData(GL.GL_ARRAY_BUFFER, position, GL.GL_STATIC_DRAW)
        GL.glVertexAttribPointer(loc, 3, GL.GL_FLOAT, False, 0, None)
        GL.glBindBuffer(GL.GL_ARRAY_BUFFER, self.buffers[1])
        GL.glBufferData(GL.GL_ARRAY_BUFFER, indices, GL.GL_STATIC_DRAW)


    def draw(self):
        # use shader, draw triangle as GL_TRIANGLE vertex array, draw array call
        GL.glUseProgram(self.shader.glid)
        GL.glBindVertexArray(self.glid)
        GL.glDrawArrays(GL.GL_TRIANGLES, 0, 18)

        # perspective

        rotation = np.radians(45.0 + glfw.get_time())
        scale_vector = np.array([1.0, 1.0, 1.0])


    def __del__(self):
        GL.glDeleteVertexArrays(1, [self.glid])
        GL.glDeleteBuffers(1, self.buffers)

File: TP1_material/c++/src/pyramid.cpp
#include "pyramid.h"

#include <iostream>
#include <GLFW/glfw3.h>
#include <glm/gtc/matrix_transform.hpp>
#include <glm/gtc/type_ptr.hpp>

Pyramid::Pyramid(Shader *shader_program) : shader_program_(shader_program->get_id()) {

// Définition des points de la pyramide
GLfloat vertex_buffer_data[] = {
    -0.5f, 0.0f,  0.5f,
     0.5f, 0.0f,  0.5f,
     0.5f, 0.0f, -0.5f,
    -0.5f, 0.0f, -0.5f,
     0.0f, 1.0f, 0.0f
};

// Définition des indices pour les faces de la pyramide
GLint indices[] = {
    0, 1, 2,
    0, 2, 3,
    0, 1, 4,
    1, 2, 4,
    2, 3, 4,
    3, 0, 4

};

    // Génération et liaison des buffers
    glGenVertexArrays(1, &VAO);
    glBindVertexArray(VAO);

    glGenBuffers(1, &VBO);
    glGenBuffers(1, &EBO);

    glEnableVertexAttribArray(0);
    glBindBuffer(GL_ARRAY_BUFFER, VBO);
    glBufferData(GL_ARRAY_BUFFER, sizeof(vertex_buffer_data), vertex_buffer_data, GL_STATIC_DRAW);

    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, EBO);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(indices), indices, GL_STATIC_DRAW);

    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 0, NULL);    
    
}

Pyramid::~Pyramid() {
    glDeleteVertexArrays(1, &VAO);
    glDeleteBuffers(1, &VBO);
}

void Pyramid::draw() {

    // Utilise le shader
    glUseProgram(this->shader_program_);
    // Lie le VAO
    glBindVertexArray(VAO);


    // Applique les transformations
    glm::mat4 rotation = glm::rotate(glm::mat4(1.0f), (float) glm::radians(50.0f * glfwGetTime()),glm::vec3(0.0f,1.0f,0.2f));
    glm::mat4 translation = glm::translate(glm::mat4(1.0f), glm::vec3(0.0f,-0.3f,-3.0f));
    glm::mat4 scale = glm::scale(glm::mat4(1.0f),glm::vec3(1.0f,1.0f,1.0f));

    // Combine les transformations
    glm::mat4 view = translation * rotation * scale;

    // Matrice de projection
    glm::mat4 projection = glm::perspective(glm::radians(45.0f),800.0f / 600.0f,0.1f,100.0f);

    // Envoie les matrices au shader
    GLint view_loc = glGetUniformLocation(this->shader_program_, "view");
    glUniformMatrix4fv(view_loc, 1, GL_FALSE, glm::value_ptr(view));
    GLint proj_loc = glGetUniformLocation(this->shader_program_, "projection");
    glUniformMatrix4fv(proj_loc, 1, GL_FALSE, glm::value_ptr(projection));

    // Dessine la pyramide
    glDrawElements(GL_TRIANGLES, 18, GL_UNSIGNED_INT, 0);
}

void Pyramid::key_handler(int key) {
    return;
}



